src/loss.py

```diff
@@ -1,2 +1,32 @@
 from abc import ABC
+import numpy as np
 
+
+class PurityFunction(ABC):
```

Comment:
If this is abstract, put it in the name:
```suggestion
class AbstractPurityFunction(ABC):
```